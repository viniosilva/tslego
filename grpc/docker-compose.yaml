version: "3.8"

services:
  app:
    build:
      context: .
      target: development
    volumes:
      - .:/usr/src/app
    ports:
      - 50051:50051
