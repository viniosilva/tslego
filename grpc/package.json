{
  "name": "tslego",
  "version": "1.0.0",
  "description": "Typescript labs",
  "scripts": {
    "build": "rimraf dist && tsc",
    "start": "node src/index.js",
    "start:dev": "ts-node src/index.ts",
    "test": "jest",
    "test:updateSnapshots": "jest --updateSnapshot",
    "test:cov": "jest --coverage",
    "test:e2e": "jest --config test/jest-e2e.json",
    "test:e2e:updateSnapshots": "jest --config test/jest-e2e.json --updateSnapshot",
    "lint": "eslint \"{src,apps,libs,test}/**/*.ts\" --fix",
    "format": "prettier --write \"src/**/*.ts\""
  },
  "jest": {
    "moduleFileExtensions": [
      "js",
      "json",
      "ts"
    ],
    "rootDir": "src",
    "testRegex": ".spec.ts$",
    "transform": {
      "^.+\\.(t|j)s$": "ts-jest"
    },
    "coverageDirectory": "../coverage",
    "testEnvironment": "node"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/viniosilva/tslego.git"
  },
  "keywords": [],
  "author": "Vinicius de Oliveira Silva",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/viniosilva/tslego/issues"
  },
  "homepage": "https://github.com/viniosilva/tslego#readme",
  "devDependencies": {
    "@types/jest": "^26.0.19",
    "@types/node": "^14.14.16",
    "@types/pino": "^6.3.4",
    "@typescript-eslint/eslint-plugin": "^4.11.0",
    "@typescript-eslint/parser": "^4.11.0",
    "eslint": "^7.16.0",
    "eslint-config-prettier": "^7.1.0",
    "eslint-config-standard": "^16.0.2",
    "eslint-plugin-import": "^2.22.1",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-promise": "^4.2.1",
    "jest": "^26.6.3",
    "prettier": "^2.2.1",
    "rimraf": "^3.0.2",
    "ts-jest": "^26.4.4",
    "ts-node": "^9.1.1",
    "typescript": "^4.1.3"
  },
  "dependencies": {
    "@grpc/proto-loader": "^0.5.5",
    "dotenv": "^8.2.0",
    "grpc": "^1.24.4",
    "install": "^0.13.0",
    "npm": "^7.3.0",
    "pino": "^6.8.0"
  }
}
