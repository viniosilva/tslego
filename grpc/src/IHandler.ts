export interface IHandler {
  proto: any;
  services: any;
}
