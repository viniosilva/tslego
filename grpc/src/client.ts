import grpc from 'grpc';
import { loadSync } from '@grpc/proto-loader';
import { apiConfig } from './common/config';

const notePackageDefinition = loadSync('protos/user.proto');
const userProto: any = grpc.loadPackageDefinition(notePackageDefinition).user;

const client = new userProto.UserService(
  `${apiConfig.host}:${apiConfig.port}`,
  grpc.credentials.createInsecure(),
);

client.getUsers(null, (err: Error, res: any) => {
  if (err) return console.error(err);
  console.log('Users Details', '\n', res);
});

client.createUser({ name: 'test' }, (err: Error, res: any) => {
  if (err) return console.error(err);
  console.log('Users creation Details', '\n', res);
});
