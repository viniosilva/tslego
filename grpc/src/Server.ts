import { Server as GrpcServer, ServerCredentials } from 'grpc';
import { Logger } from 'pino';
import { IHandler } from './IHandler';

export class Server {
  private readonly server = new GrpcServer();

  constructor(
    private readonly host: string,
    private readonly port: number,
    handlers: IHandler[],
    private readonly logger: Logger,
  ) {
    handlers.forEach((handler: IHandler) => {
      this.server.addService(handler.proto.service, handler.services);
    });
  }

  start() {
    const uri = `${this.host}:${this.port}`;
    this.server.bind(uri, ServerCredentials.createInsecure());
    this.server.start();

    this.logger.info(`App listening on [${this.host}:${this.port}]`);
  }
}
