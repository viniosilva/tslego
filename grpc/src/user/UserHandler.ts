import { loadSync } from '@grpc/proto-loader';
import {
  loadPackageDefinition,
  sendUnaryData,
  ServerUnaryCall,
  ServiceError,
  status,
} from 'grpc';
import { Logger } from 'pino';
import { Response } from '../dto/Response';
import { IHandler } from '../IHandler';
import { User, UserRequest, UserRequestId } from './dto/User';
import { NotFoundException } from './exception/NotFoundException';
import { UserService } from './UserService';

export class UserHandler implements IHandler {
  readonly proto: any;
  readonly services: any;

  constructor(
    private readonly userService: UserService,
    private readonly logger: Logger,
  ) {
    const packageDefinition = loadSync('protos/user.proto');
    const proto: any = loadPackageDefinition(packageDefinition).user;
    this.proto = proto.UserService;

    this.services = {
      getUsers: this.getUsers.bind(this),
      getUserById: this.getUserById.bind(this),
      createUser: this.createUser.bind(this),
      updateUserById: this.updateUserById.bind(this),
      removeUserById: this.removeUserById.bind(this),
    };
  }

  getUsers(
    call: ServerUnaryCall<null>,
    callback: sendUnaryData<Response>,
  ): void {
    try {
      this.logRequest(call.request, 'getUsers');
      callback(null, { data: this.userService.getUsers() });
    } catch (err) {
      this.errorHandler(err, callback);
    }
  }

  getUserById(
    call: ServerUnaryCall<UserRequestId>,
    callback: sendUnaryData<Response>,
  ): void {
    try {
      this.logRequest(call.request, 'getUserById');
      callback(null, {
        data: this.userService.getUserById(call.request.id),
      });
    } catch (err) {
      this.errorHandler(err, callback);
    }
  }

  createUser(
    call: ServerUnaryCall<UserRequest>,
    callback: sendUnaryData<Response>,
  ): void {
    try {
      this.logRequest(call.request, 'createUser');
      callback(null, {
        data: this.userService.createUser(call.request),
      });
    } catch (err) {
      this.errorHandler(err, callback);
    }
  }

  updateUserById(
    call: ServerUnaryCall<User>,
    callback: sendUnaryData<Response>,
  ): void {
    try {
      this.logRequest(call.request, 'updateUserById');
      callback(null, {
        data: this.userService.updateUserById(call.request.id, call.request),
      });
    } catch (err) {
      this.errorHandler(err, callback);
    }
  }

  removeUserById(
    call: ServerUnaryCall<UserRequestId>,
    callback: sendUnaryData<Response>,
  ): void {
    try {
      this.logRequest(call.request, 'removeUserById');
      this.userService.removeUserById(call.request.id);
      callback(null, { message: 'OK' });
    } catch (err) {
      this.errorHandler(err, callback);
    }
  }

  private logRequest(request: any, service: string): void {
    this.logger.child({ service, request }).info('Request completed');
  }

  private errorHandler(err: Error, callback: sendUnaryData<Response>) {
    if (err instanceof NotFoundException) {
      callback(
        { code: status.NOT_FOUND, message: err.message } as ServiceError,
        null,
      );
    } else {
      this.logger.child({ stackError: err.stack }).error(err.message);
      callback(
        {
          code: status.INTERNAL,
          message: 'Internal server error',
        } as ServiceError,
        null,
      );
    }
  }
}
