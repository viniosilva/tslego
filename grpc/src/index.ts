import { apiConfig } from './common/config';
import { logger } from './common/logger';
import { Server } from './Server';
import { userHandler } from './user';

const handlers = [userHandler];

const server = new Server(apiConfig.host, apiConfig.port, handlers, logger);
server.start();
