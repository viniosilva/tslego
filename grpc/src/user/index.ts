import { logger } from '../common/logger';
import { UserHandler } from './UserHandler';
import { UserService } from './UserService';

const userService = new UserService();

export const userHandler = new UserHandler(userService, logger);
