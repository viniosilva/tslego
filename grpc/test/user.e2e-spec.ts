import { credentials } from 'grpc';
import { userHandler } from '../src/user';
import { apiConfig } from '../src/common/config';

describe('User E2E', () => {
  const client = new userHandler.proto(
    `${apiConfig.host}:${apiConfig.port}`,
    credentials.createInsecure(),
  );

  describe('getUsers', () => {
    it('should return empty user list', (done) => {
      client.getUsers(null, (err: Error, res: any) => {
        if (err) throw err;
        expect(res).toMatchSnapshot();
        done();
      });
    });

    it('should return user list', (done) => {
      client.createUser({ name: 'test' }, (err: Error, _res: any) => {
        if (err) throw err;

        client.getUsers(null, (err: Error, res: any) => {
          if (err) throw err;
          expect(res).toMatchSnapshot();
          done();
        });
      });
    });
  });

  describe('getUserById', () => {
    it('should return user', (done) => {
      client.createUser({ name: 'test' }, (err: Error, res: any) => {
        if (err) throw err;

        client.getUserById({ id: res.data.id }, (err: Error, res: any) => {
          if (err) throw err;
          expect(res).toMatchSnapshot();
          done();
        });
      });
    });

    it('should throw not found exception', (done) => {
      client.getUserById({ id: 0 }, (err: Error, _res: any) => {
        expect(err).toMatchSnapshot();
        done();
      });
    });
  });

  describe('createUser', () => {
    it('should return new user', (done) => {
      client.createUser({ name: 'test' }, (err: Error, res: any) => {
        if (err) throw err;
        expect(res).toMatchSnapshot();
        done();
      });
    });

    // it('should throw bad request exception', () => {});
  });

  describe('updateUserById', () => {
    it('should return updated user', (done) => {
      client.createUser({ name: 'test' }, (err: Error, res: any) => {
        if (err) throw err;

        client.updateUserById(
          { id: res.data.id, name: 'update test' },
          (err: Error, res: any) => {
            if (err) throw err;
            expect(res).toMatchSnapshot();
            done();
          },
        );
      });
    });

    // it('should throw bad request exception', () => {});

    it('should throw not found exception', (done) => {
      client.updateUserById(
        { id: 0, name: 'update test' },
        (err: Error, res: any) => {
          expect(err).toMatchSnapshot();
          done();
        },
      );
    });
  });

  describe('removeUserById', () => {
    it('should remove user', (done) => {
      client.createUser({ name: 'test' }, (err: Error, res: any) => {
        if (err) throw err;

        client.removeUserById({ id: res.data.id }, (err: Error, res: any) => {
          if (err) throw err;
          expect(res).toMatchSnapshot();
          done();
        });
      });
    });
  });
});
