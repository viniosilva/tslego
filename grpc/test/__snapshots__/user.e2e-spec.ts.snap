// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User E2E createUser should return new user 1`] = `
Object {
  "data": Object {
    "id": 3,
    "name": "test",
  },
}
`;

exports[`User E2E getUserById should return user 1`] = `
Object {
  "data": Object {
    "id": 2,
    "name": "test",
  },
}
`;

exports[`User E2E getUserById should throw not found exception 1`] = `[Error: 5 NOT_FOUND: User not found]`;

exports[`User E2E getUsers should return empty user list 1`] = `Object {}`;

exports[`User E2E getUsers should return user list 1`] = `
Object {
  "data": Array [
    Object {
      "id": 1,
      "name": "test",
    },
  ],
}
`;

exports[`User E2E removeUserById should remove user 1`] = `
Object {
  "message": "OK",
}
`;

exports[`User E2E updateUserById should return updated user 1`] = `
Object {
  "data": Object {
    "id": 4,
    "name": "update test",
  },
}
`;

exports[`User E2E updateUserById should throw not found exception 1`] = `[Error: 5 NOT_FOUND: User not found]`;
